### PR TITLE
runtime-hooks: add auto register hooks

### LINKED ISSUE #496: [proposal] register hook plugins of koordlet to runtime proxy auto at start
**What is your proposal**:

**Why is this needed**:

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by songtao98**
### Ⅰ. Describe what this PR does
Add auto register runtime hooks of koordlet to runtime proxy.

### Ⅱ. Does this pull request fix one issue?
fixes #496 

### Ⅳ. Special notes for reviews
Unit tests and code format will be pushed later.
### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/runtimehooks/runtimehooks.go [hunk 1]
```diff
@@ -87,3 +95,24 @@ func registerPlugins() {
 		klog.Infof("runtime hook plugin %s enable %v", hookFeature, enabled)
 	}
 }
+
+const (
+	defaultRuntimeHookConfigFilePath string = "/etc/runtime/hookserver.d/koordlet.json"
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/runtimehooks.go:100]: make defaultRuntimeHookConfigFilePath as an arg, besides, remember to record an issue on charts repo to update the koordlet yaml.

Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/runtimehooks.go:100]: path inside container and hosts are not same

File: pkg/koordlet/runtimehooks/runtimehooks.go [hunk 2]
```diff
@@ -17,7 +17,12 @@ limitations under the License.
 package runtimehooks
 
 import (
+	"encoding/json"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/hooks"
+	"github.com/koordinator-sh/koordinator/pkg/runtimeproxy/config"
 	"k8s.io/klog/v2"
+	"os"
+	"path/filepath"
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/runtimehooks.go:25]: sort import

File: pkg/koordlet/runtimehooks/runtimehooks.go [hunk 3]
```diff
@@ -47,14 +52,17 @@ func (r *runtimeHook) Run(stopCh <-chan struct{}) error {
 	if err := r.reconciler.Run(stopCh); err != nil {
 		return err
 	}
+	if err := registerPluginsToProxy(); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/runtimehooks.go:55]: define Option in runtimeHook struct
r.registerPluginsToProxy

File: pkg/koordlet/runtimehooks/runtimehooks.go [hunk 4]
```diff
@@ -87,3 +95,24 @@ func registerPlugins() {
...
+	hookConfig := &config.RuntimeHookConfig{
+		RemoteEndpoint: serverOptions.Address,
+		FailurePolicy:  serverOptions.FailurePolicy,
+		RuntimeHooks:   hooks.GetStages(),
+	}
+
+	configData, _ := json.Marshal(hookConfig)
+
+	// todo: confirm perm value
+	if err := os.WriteFile(filepath.Join(defaultRuntimeHookConfigFilePath), configData, 0666); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/runtimehooks.go:114]: use util.*

File: pkg/koordlet/runtimehooks/runtimehooks.go (hunk 3 above)
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/runtimehooks.go:55]: r.server.registerPluginServerToProxy

**Comment by zwzhang0107** [comment]
make lint
make test
commit -asm --amend "xxx"

File: pkg/koordlet/runtimehooks/config.go [hunk 5]
```diff
@@ -49,22 +49,25 @@ var (
 )
 
 type Config struct {
-	RuntimeHooksNetwork string
-	RuntimeHooksAddr    string
-	FeatureGates        map[string]bool
+	RuntimeHooksNetwork       string
+	RuntimeHooksAddr          string
+	RuntimeHooksFailurePolicy string
```
Comment (jasonliu747) [inline pkg/koordlet/runtimehooks/config.go:54]: you probably need to validate `RuntimeHooksFailurePolicy`

File: pkg/koordlet/runtimehooks/proxyserver/server.go [hunk 6]
```diff
@@ -18,6 +18,7 @@ package proxyserver
 
 import (
 	"fmt"
+	"github.com/koordinator-sh/koordinator/pkg/runtimeproxy/config"
```
Comment (jasonliu747) [inline pkg/koordlet/runtimehooks/proxyserver/server.go:21]: please sort imports

File: pkg/runtimeproxy/config/config.go [hunk 7]
```diff
@@ -27,6 +27,17 @@ const (
 	PolicyIgnore FailurePolicyType = "Ignore"
 )
 
+func GetPolicyType(typeString string) FailurePolicyType {
```
Comment (eahydra) [inline pkg/runtimeproxy/config/config.go:30]: GetFailurePolicyType?

File: pkg/koordlet/runtimehooks/proxyserver/server.go [hunk 8]
```diff
@@ -17,26 +17,34 @@ limitations under the License.
...
+	Address        string
+	FailurePolicy  config.FailurePolicyType
+	ConfigFilePath string
 }
 
 type Server interface {
 	Setup() error
 	Start() error
 	Stop()
+	RegisterPluginServerToProxy() error
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/proxyserver/server.go:47]: how about `Register()` or `RegisterToProxy`

File: pkg/koordlet/runtimehooks/runtimehooks.go [hunk 9]
```diff
@@ -47,14 +48,23 @@ func (r *runtimeHook) Run(stopCh <-chan struct{}) error {
...
 	klog.Infof("runtime hook is stopped")
 	return nil
 }
 
 func NewRuntimeHook(si statesinformer.StatesInformer, cfg *Config) (RuntimeHook, error) {
-	s, err := proxyserver.NewServer(proxyserver.Options{Network: cfg.RuntimeHooksNetwork, Address: cfg.RuntimeHooksAddr})
+	newServerOptions := proxyserver.Options{
+		Network:        cfg.RuntimeHooksNetwork,
+		Address:        cfg.RuntimeHooksAddr,
+		FailurePolicy:  config.GetFailurePolicyType(cfg.RuntimeHooksFailurePolicy),
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/runtimehooks.go:64]: return error if config is illegal，make koordlet start failed

File: pkg/koordlet/runtimehooks/config.go [hunk 10]
```diff
@@ -29,8 +29,9 @@ import (
 )
 
 const (
-	GroupIdentity   featuregate.Feature = "GroupIdentity"
-	CPUSetAllocator featuregate.Feature = "CPUSetAllocator"
+	GroupIdentity                    featuregate.Feature = "GroupIdentity"
+	CPUSetAllocator                  featuregate.Feature = "CPUSetAllocator"
+	defaultRuntimeHookConfigFilePath string              = "/etc/runtime/hookserver.d"
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/config.go:34]: set default path same as helm yaml，such as /host-etc-hookserver/

File: pkg/koordlet/runtimehooks/proxyserver/server.go [hunk 11]
```diff
@@ -67,6 +75,19 @@ func (s *server) Stop() {
...
 }
 
+func (s *server) RegisterPluginServerToProxy() error {
+	hookConfig := &config.RuntimeHookConfig{
+		RemoteEndpoint: s.options.Address,
+		FailurePolicy:  s.options.FailurePolicy,
+		RuntimeHooks:   hooks.GetStages(),
+	}
+	configData, _ := json.Marshal(hookConfig)
+	if err := ioutil.WriteFile(filepath.Join(s.options.ConfigFilePath, "koordlet.json"), configData, 0644); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/proxyserver/server.go:85]: i think koordinator has defined some io utils can be reused

File: pkg/koordlet/runtimehooks/proxyserver/server.go (hunk 8 above)
Comment (songtao98) [reply pkg/koordlet/runtimehooks/proxyserver/server.go:47]: Sure, `Register()` would be better

**Review by jasonliu747 — APPROVED** [review]
/lgtm

File: pkg/koordlet/runtimehooks/proxyserver/server.go [hunk 12]
```diff
@@ -67,6 +75,19 @@ func (s *server) Stop() {
 	s.server.Stop()
 }
 
+func (s *server) Register() error {
+	hookConfig := &config.RuntimeHookConfig{
+		RemoteEndpoint: s.options.Address,
+		FailurePolicy:  s.options.FailurePolicy,
+		RuntimeHooks:   hooks.GetStages(),
```
Comment (hormes) [inline pkg/koordlet/runtimehooks/proxyserver/server.go:82]: 就是这里，hooks.GetStages 是所以需要注册的 hook 点（需求），但这里是否注册，还差了一个一个管理接口，比如忽略掉 PreStopContainer。

**Pull request merged**